use whole-program code-generation for all backends

zerbina opened the pull request:
## Summary

Change the C and JavaScript backends to use whole-program code-
generation in the same way that the VM and IC backends do. In short,
this means that code generation is now only run after *all* modules part
of the program were semantically analyzed.

This brings the architecture of the C and JS backends closer to that of
the VM backend, and is the first step towards unifying the backend
processing.

The final goal is to have all backends process code in the same way,
with as much as possible of the pre-processing currently performed by
the code generators being moved to a shared, backend-agnostic layer.

## Details

Generalize the `ModuleGraph` pass that the VM backend used for gathering
the AST of each alive module, and move it into its own module. Compared
to the original implementation, the generalized pass:
* doesn't drop declarative nodes (the VM backend doesn't care about
  them, but the C and JS backends do)
* remembers the `IdGenerator` associated with each module
* doesn't introduce its own module order (that's left to the backend)
* uses the more descriptive `SeqMap` instead of a raw `seq`

`vmbackend` is adjusted to work with the generalized collection pass:
the module list produced by the pass is translated into the structure
that the rest of the VM backend still expects, but a `Store` with a
dedicated ID type is now used instead of a raw `seq` (preparing for
future improvements).

Using the same naming scheme as `vmbackend`, the modules `cbackend`
and `jsbackend` are introduced -- they implement the code-generation
orchestrators for the C and JavaScript backends, respectively. The
`passes` integration is removed from `cgen` and `jsgen`, as invoking the
code generators is now the responsibility of the orchestrators.

The new orchestrators take the module list produced by the collector
pass and generate the code for it (with `jsbackend` also writing the
output to disk already). They're very basic at the moment, but in the
future will take on similar responsibilities as `vmbackend` does
for the VM backend (e.g., dead-code elimination, running `transf`,
etc.).

Since forwarded procedures don't reach the code generators anymore, the
special handling for them is removed.

### Known Issues

Changes to options performed by `.push` having an effect on top-level
code relied on semantic analysis and code generation happening in a
pipelined manner.

Since this is no longer the case, disabling or enabling checks for top-
level code stops working for now, but the plan is to bring this feature
back in the future.

zerbina committed:
containers: extend the APIs

Extend the set of routines for interacting with `Store` and `SeqMap`.

zerbina committed:
move the 'collect' pass to a dedicated module

zerbina committed:
make `vmbackend` compile again

The `ModuleList` used by the collector pass is now no longer re-used in
the VM backend, but instead translated to a custom representation (see
`produceModules`).

In addition, the list with the top-level AST is now discarded after it's
passed to code generation, which should reduce memory pressure, and thus
peak memory usage, quite a bit.

zerbina committed:
collectors: track a module's `IdGenerator`

The upcoming C code-generation orchestrator needs access to the
`IdGenerator` for each module. They're now stored with `FullModule`,
allowing the orchestrator to retrieve them later.

`vmbackend` is adjusted to make use of the module's `IdGenerator`, which
is more precise than using the `ModuleGraph`'s one. Ideally, the backend
should not introduce any new symbol and type instances (which is what
the `IdGenerator`s are required for), but `transf` currently
necessitates that.

zerbina committed:
add a C code-generation orchestrator

Much like how it works for the VM backend, the C backend now also uses
an orchestrator (the new `cbackend` module) that invokes the code
generator.

The `passes` integration (`myOpen`, `myProcess`, and `myClose`) is
removed from `cgen`. Setting up the extra header backend module (used
for C header generation) is now the responsibility of the orchestrator.

Similar to `vmbackend`, the new `cbackend` also operates on the
semantically analyzed AST of the whole program.

zerbina committed:
collectors: simpler module list management

Store the modules in a `SeqMap` with their position as the key. This
conveys intention better and leaves it to the consumer to use a more
fitting structure/representation.

For the internal representation, `vmbackend` uses a `Store` for the
list of modules, which makes the intention more clear and querying the
list a bit more ergonomic.

zerbina committed:
cgen: remove handling of forwarded procedures

Since code generation now only takes place *after* the whole program was
semantically analyzed, forwarded procedures and unresolved borrows no
longer reach there.

zerbina committed:
add a JS code-generation orchestrator

Very similar to the introduction of an orchestrator for the C code-
generator, but for the JS code-generator.

The `passes` integration is removed form `jsgen`, and writing the module
to disk and generating the source map moved to the orchestrator.

zerbina committed:
jsgen: remove `inSystem`

Disabling stack-traces for the system module and all modules it imports
differs from what the C code-generator does (stack-traces are only
disabled for the `system` module there, but not for the ones it
imports).

The required access to `PGlobals` via `g.backend` is also a small
problem for upcoming `jsgen` refactoring.

zerbina committed:
pass the `ModuleList` to the `generateCode` routines

Instead of extracting the `ModuleList` as part of `generateCode`, it is
now passed in as a `sink` parameter. This:
* makes it easier to move away from storing the `ModuleList` as part of
  the `ModuleGraph`
* moves the mutation to the callsite
* allows for implementing the memory consumption optimization employed
  by `vmbackend` (which is eventually going to be used for the other
  backends too) in a much cleaner way

zerbina committed:
tests: disable parts of the `toverflw.nim` test

With code generation now always happening after all semantic analysis is
done, option changes applied via the `push` and `pop` pragmas no longer
apply to top-level statements.

As a temporary solution, the feature could be made to work by processing
`nkPragma` nodes in the orchestrators, but the planned upcoming changes
would render this approach unusable again.

zerbina committed:
jsgen: remove handling of forwarded procedures

zerbina committed:
collectors: don't drop declarative nodes

The pass should only collect the statements into a list, and not
introduce its own decision making.

saem commented on compiler/backend/collectors.nim:
```diff
@@ -0,0 +1,82 @@
+## Implements the "collect" pass. This pass gathers the full AST for each
+## module into a single structure, which is then meant to be consumed by the
+## code-generation orchestrators.
```
```suggestion
## code-generation orchestrators (`cbackend`, `jsbackend`, etc).
```

Orchestrator doesn't appear in any identifiers so I figured a bit of clarity here might send folks new to the code in the right direction.

saem commented on compiler/modules/modulegraphs.nim:
```diff
@@ -108,7 +108,9 @@ type
-    backend*: RootRef # minor hack so that a backend can extend this easily
+    backend*: RootRef # XXX: having this field is a hack, but it's still
+                      #      required by the current ``passes`` design. Remove
+                      #      the ``passes`` design is phased out
```
```suggestion
                      #      once the ``passes`` design is phased out
```

saem commented on tests/overflw/toverflw.nim:
```diff
@@ -1,24 +1,31 @@
-doAssert(overflowDetected)
+# XXX: overflow checks (and other checks) cannot be disabled for module-level
+#      code at the moment
+doAssert(overflowDetected == false)
```
```suggestion
doAssert(overflowDetected == false), "re-enable this test, module-level code bounds checking now works"
```

saem reviewed (approved):
Looks good.

I left a few minor suggestions mostly to do with a typo here or there in comments.

~I saw a few in the PR desc, I'll tackle those after I take a walk.~

saem commented:
I was going to suggest it but things are too different right now, but we can probably get away with a single `backend` module, that has a single public routine, something like: `generateCode*(g: ModuleGraph, mlist: sink ModuleList, backend: TBackend)`, this would evolve over time, likely become an iterator producing progress events. But that's all nice to have and that consolidation doesn't buy us enough as things are still not similar enough.

zerbina commented:
> I was going to suggest it but things are too different right now, but we can probably get away with a single `backend` module

Yep, that's what I initially tried as part of #550, but the differences in processing were too large for it to make sense at this time. With that said, having a single entry point into the backend *is* roughly the direction I'm going in.

zerbina committed:
address review comments

Thank you for the review, @saem .

zerbina committed:
cbackend: clarify the "close module" comment

The comment was misleading, as a backend module is not closed for
writing after the call to `finalCodegenActions`.

saem commented:
/merge
